### ISSUE TITLE
[openssl] Build error on Android arm/armv6/x86 of OpenSSL 3.2.0 (regression after update from OpenSSL 3.1.4)

### ISSUE THREAD

**Issue by AenBleidd** [microsoft/vcpkg#35792]
armv6:
```
/home/runner/work/boinc/boinc/3rdParty/android/android-ndk-r15c/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot open crtbegin_dynamic.o: No such file or directory
/home/runner/work/boinc/boinc/3rdParty/android/android-ndk-r15c/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: fatal error: /home/runner/work/boinc/boinc/3rdParty/android/android-ndk-r15c/platforms/android-16/arch-arm: pread failed: Is a directory
```
arm:
```
../src/nssl-3.2.0-06d693f20a.clean/crypto/bio/bss_dgram.c:1371:11: warning: implicit declaration of function 'sendmmsg' is invalid in C99 [-Wimplicit-function-declaration]
    ret = sendmmsg(b->num, mh, num_msg, sysflags);
          ^
../src/nssl-3.2.0-06d693f20a.clean/crypto/bio/bss_dgram.c:1564:11: warning: implicit declaration of function 'recvmmsg' is invalid in C99 [-Wimplicit-function-declaration]
    ret = recvmmsg(b->num, mh, num_msg, sysflags, NULL);
          ^
2 warnings generated.
ld: error: undefined symbol: sendmmsg
>>> referenced by bss_dgram.c:1371 (../src/nssl-3.2.0-06d693f20a.clean/crypto/bio/bss_dgram.c:1371)
>>>               libcrypto-lib-bss_dgram.o:(dgram_sendmmsg) in archive libcrypto.a
>>> did you mean: sendmsg
>>> defined in: /home/runner/work/boinc/boinc/3rdParty/android/android-ndk-r22b/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/16/libc.so

ld: error: undefined symbol: recvmmsg
>>> referenced by bss_dgram.c:1564 (../src/nssl-3.2.0-06d693f20a.clean/crypto/bio/bss_dgram.c:1564)
>>>               libcrypto-lib-bss_dgram.o:(dgram_recvmmsg) in archive libcrypto.a
>>> did you mean: recvmsg
>>> defined in: /home/runner/work/boinc/boinc/3rdParty/android/android-ndk-r22b/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/16/libc.so
```
x86:
```
../src/nssl-3.2.0-06d693f20a.clean/crypto/bio/bss_dgram.c:1371:11: warning: implicit declaration of function 'sendmmsg' is invalid in C99 [-Wimplicit-function-declaration]
    ret = sendmmsg(b->num, mh, num_msg, sysflags);
          ^
../src/nssl-3.2.0-06d693f20a.clean/crypto/bio/bss_dgram.c:1564:11: warning: implicit declaration of function 'recvmmsg' is invalid in C99 [-Wimplicit-function-declaration]
    ret = recvmmsg(b->num, mh, num_msg, sysflags, NULL);
          ^
2 warnings generated.
../src/nssl-3.2.0-06d693f20a.clean/crypto/threads_pthread.c:228:16: warning: misaligned atomic operation may incur significant performance penalty [-Watomic-alignment]
        *ret = __atomic_or_fetch(val, op, __ATOMIC_ACQ_REL);
               ^
../src/nssl-3.2.0-06d693f20a.clean/crypto/threads_pthread.c:253:9: warning: misaligned atomic operation may incur significant performance penalty [-Watomic-alignment]
        __atomic_load(val, ret, __ATOMIC_ACQUIRE);
        ^
2 warnings generated.
ld: error: undefined symbol: sendmmsg
>>> referenced by bss_dgram.c:1371 (../src/nssl-3.2.0-06d693f20a.clean/crypto/bio/bss_dgram.c:1371)
>>>               libcrypto-lib-bss_dgram.o:(dgram_sendmmsg) in archive libcrypto.a
>>> did you mean: sendmsg
>>> defined in: /home/runner/work/boinc/boinc/3rdParty/android/android-ndk-r22b/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/16/libc.so

ld: error: undefined symbol: recvmmsg
>>> referenced by bss_dgram.c:1564 (../src/nssl-3.2.0-06d693f20a.clean/crypto/bio/bss_dgram.c:1564)
>>>               libcrypto-lib-bss_dgram.o:(dgram_recvmmsg) in archive libcrypto.a
>>> did you mean: recvmsg
>>> defined in: /home/runner/work/boinc/boinc/3rdParty/android/android-ndk-r22b/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/16/libc.so
```

For all three triplets I have one difference from the official triplets:
```
set(VCPKG_CMAKE_SYSTEM_VERSION android-16)
```

**Comment by AenBleidd** [comment]
Reference with the failed build: https://github.com/BOINC/boinc/actions/runs/7279872486

**Comment by dg0yt** [comment]
Probably not a vcpkg port bug, but caused upstream. It may need more than android-16 now, or better detection of system capabilities if there is an alternative implementation.

**Comment by JonLiu1993** [comment]
@AenBleidd, Thanks for posting this issue, I haven't reproduced your problem locally. Here are the steps I took to reproduce the issue:
git blame -l versions/o-/openssl.json:
````
795111ff31e2b9f87fa4fecf689d4ca16a616188 versions/o-/openssl.json      (Evgeny Karpov              2023-11-05 18:34:01 +0100  10)       "version": "3.1.4",
795111ff31e2b9f87fa4fecf689d4ca16a616188 versions/o-/openssl.json      (Evgeny Karpov              2023-11-05 18:34:01 +0100  11)       "port-version": 1
795111ff31e2b9f87fa4fecf689d4ca16a616188 versions/o-/openssl.json      (Evgeny Karpov              2023-11-05 18:34:01 +0100  12)     },
98a562a04cd03728f399e79e1b37bcccb5a69b37 versions/o-/openssl.json      (Frank                      2023-10-28 02:30:01 +0800  13)     {
98a562a04cd03728f399e79e1b37bcccb5a69b37 versions/o-/openssl.json      (Frank                      2023-10-28 02:30:01 +0800  14)   
````
./vcpkg install --triplet=x86-windows:
````
Building openssl:x86-windows...
-- Installing port from location: F:\test\vcpkg\buildtrees\versioning_\versions\openssl\5ee74387a1dd212dc9754c614809ff86bd0638fa
-- Using cached openssl-openssl-openssl-3.1.4.tar.gz.
-- Cleaning sources at F:/test/vcpkg/buildtrees/openssl/src/nssl-3.1.4-c5488d00d9.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source F:/test/vcpkg/downloads/openssl-openssl-openssl-3.1.4.tar.gz
-- Applying patch disable-apps.patch
-- Applying patch disable-install-docs.patch
-- Applying patch script-prefix.patch
-- Applying patch windows/install-layout.patch
-- Applying patch windows/install-pdbs.patch
-- Applying patch unix/android-cc.patch
-- Applying patch unix/move-openssldir.patch
-- Applying patch unix/no-empty-dirs.patch
-- Applying patch unix/no-static-libs-for-shared.patch
-- Using source at F:/test/vcpkg/buildtrees/openssl/src/nssl-3.1.4-c5488d00d9.clean
-- Found external ninja('1.11.0').
-- Getting CMake variables for x86-windows
-- Getting CMake variables for x86-windows
-- Prerunning x86-windows-dbg
-- Building x86-windows-dbg
-- Prerunning x86-windows-rel
-- Building x86-windows-rel
-- Fixing pkgconfig file: F:/test/vcpkg/packages/openssl_x86-windows/lib/pkgconfig/libcrypto.pc
-- Fixing pkgconfig file: F:/test/vcpkg/packages/openssl_x86-windows/lib/pkgconfig/libssl.pc
-- Fixing pkgconfig file: F:/test/vcpkg/packages/openssl_x86-windows/lib/pkgconfig/openssl.pc
-- Using cached mingw-w64-x86_64-pkgconf-1~2.1.0-1-any.pkg.tar.zst.
-- Using cached msys2-msys2-runtime-3.4.9-3-x86_64.pkg.tar.zst.
-- Using msys root at F:/test/vcpkg/downloads/tools/msys2/023cdb3ca06f77f2
-- Fixing pkgconfig file: F:/test/vcpkg/packages/openssl_x86-windows/debug/lib/pkgconfig/libcrypto.pc
-- Fixing pkgconfig file: F:/test/vcpkg/packages/openssl_x86-windows/debug/lib/pkgconfig/libssl.pc
-- Fixing pkgconfig file: F:/test/vcpkg/packages/openssl_x86-windows/debug/lib/pkgconfig/openssl.pc
-- Installing: F:/test/vcpkg/packages/openssl_x86-windows/share/openssl/usage
-- Installing: F:/test/vcpkg/packages/openssl_x86-windows/share/openssl/copyright
-- Performing post-build validation
Stored binaries in 1 destinations in 2.2 s.
Elapsed time to handle openssl:x86-windows: 5.6 min
Total install time: 5.6 min
The package openssl is compatible with built-in CMake targets:

    find_package(OpenSSL REQUIRED)
    target_link_libraries(main PRIVATE OpenSSL::SSL OpenSSL::Crypto)
````
vcpkg.json:
````
{
  "name": "test",
  "version-string": "0.0.1",
  "dependencies": [
    "openssl"
  ],
  "builtin-baseline": "795111ff31e2b9f87fa4fecf689d4ca16a616188"
}
````

**Comment by AenBleidd** [comment]
Looks like it depends on installed NDK as well.
Nevermind, I know how to fix it (at least for x86 and arm), I'll create a PR as soon as I have working solution

**Comment by AenBleidd** [comment]
Ok, good news. I have fixed all these three issues.
For x86/arm I added next lines to the triplets:
```
set(VCPKG_CXX_FLAGS -DNO_RECVMMSG)
set(VCPKG_C_FLAGS -DNO_RECVMMSG)
```
Looks like OpenSSL doesn't detect availability of the libc functions (I'm not very familiar with Perl and OpenSSL configure file, so if anyone points me to the place where they check for the functions - I'll happily add the patch to the OpenSSL upstream).

For armv6 the situation if more interesting: they have added a `quicserver` tool that looks like not needed at all when building an OpenSSL library, and used only for testing. They have no separate flag to disable building this particular utility, so I have to patched that in the next PR: #35838
